L1Trigger/TrackFindingTracklet/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,261 @@
+    // helper class to extract structured data from TTDTC::Frames
+    const DataFormats* dataFormats_;
+    // helper class to encode layer
+    const LayerEncoding* layerEncoding_;
+    // helper class to assign tracks to channel
+    TrackBuilderChannel* trackBuilderChannel_;
+    //
+    bool enableTruncation_;
+  };
+
+  ProducerKFin::ProducerKFin(const ParameterSet& iConfig) :
+    iConfig_(iConfig)
+  {
+    const InputTag& inputTag = iConfig.getParameter<InputTag>("InputTag");
+    const string& branchAcceptedStubs = iConfig.getParameter<string>("BranchAcceptedStubs");
+    const string& branchAcceptedTracks = iConfig.getParameter<string>("BranchAcceptedTracks");
+    const string& branchLostStubs = iConfig.getParameter<string>("BranchLostStubs");
+    const string& branchLostTracks = iConfig.getParameter<string>("BranchLostTracks");
+    // book in- and output ED products
+    edGetTokenTTTracks_ = consumes<TTTracks>(inputTag);
+    edPutTokenAcceptedStubs_ = produces<TTDTC::Streams>(branchAcceptedStubs);
+    edPutTokenAcceptedTracks_ = produces<StreamsTrack>(branchAcceptedTracks);
+    edPutTokenLostStubs_ = produces<TTDTC::Streams>(branchLostStubs);
+    edPutTokenLostTracks_ = produces<StreamsTrack>(branchLostTracks);
+    // book ES products
+    esGetTokenSetup_ = esConsumes<Setup, SetupRcd, Transition::BeginRun>();
+    esGetTokenDataFormats_ = esConsumes<DataFormats, DataFormatsRcd, Transition::BeginRun>();
+    esGetTokenLayerEncoding_ = esConsumes<LayerEncoding, LayerEncodingRcd, Transition::BeginRun>();
+    esGetTokenTrackBuilderChannel_ = esConsumes<TrackBuilderChannel, TrackBuilderChannelRcd, Transition::BeginRun>();
+    // initial ES products
+    setup_ = nullptr;
+    dataFormats_ = nullptr;
+    layerEncoding_ = nullptr;
+    trackBuilderChannel_ = nullptr;
+    //
+    enableTruncation_ = iConfig.getParameter<bool>("EnableTruncation");
+  }
+
+  void ProducerKFin::beginRun(const Run& iRun, const EventSetup& iSetup) {
+    // helper class to store configurations
+    setup_ = &iSetup.getData(esGetTokenSetup_);
+    if (!setup_->configurationSupported())
+      return;
+    // check process history if desired
+    if (iConfig_.getParameter<bool>("CheckHistory"))
+      setup_->checkHistory(iRun.processHistory());
+    // helper class to extract structured data from TTDTC::Frames
+    dataFormats_ = &iSetup.getData(esGetTokenDataFormats_);
+    // helper class to encode layer
+    layerEncoding_ = &iSetup.getData(esGetTokenLayerEncoding_);
+    // helper class to assign tracks to channel
+    trackBuilderChannel_ = const_cast<TrackBuilderChannel*>(&iSetup.getData(esGetTokenTrackBuilderChannel_));
+  }
+
+  void ProducerKFin::produce(Event& iEvent, const EventSetup& iSetup) {
+    auto toFrameStub = [](StubKFin* stub) { return stub->frame(); };
+    auto toFrameTrack = [](const TrackKFin& track){ return track.frame(); };
+    const DataFormat& dfCot = dataFormats_->format(Variable::cot, Process::sf);
```

Comment:
Add comment: "Get utility with digi & float representations of each variable."

Reply:
done

Reply:
not done?

Reply:
I wrote // dataformat used for track cotTheta wrt eta sector centre